Update Learn Sensu lesson in sandbox (#1833)
sensu/sensu-docs: content/sensu-go/5.14/getting-started/learn-sensu.md

git clone https://github.com/sensu/sandbox && cd sandbox/sensu-go
{{< /highlight >}}

_**NOTE**: If you've cloned the sandbox repository before, run `git pull https://github.com/sensu/sandbox && cd sandbox/sensu-go` instead._

> 

Author: @rgeniesse does this ^ cover it?

{{< /highlight >}}

The Learn Sensu sandbox is a CentOS 7 virtual machine pre-installed with Sensu, InfluxDB, and Grafana.

> It's intended for you to use as a learning tool--we do not recommend using it in a production installation. To install Sensu in production, use the [installation guide](../../installation/install-sensu) instead.

Reviewer: minor nit (and maybe the format isn't translating here), but could we please use an em dash with a space on either end? e.g., "...a learning tool — we do not recommend..."

> It's intended for you to use as a learning tool--we do not recommend using it in a production installation. To install Sensu in production, use the [installation guide](../../installation/install-sensu) instead.

Author: ```suggestion
It's intended for you to use as a learning tool &mdash; we do not recommend using it in a production installation. To install Sensu in production, use the [installation guide](../../installation/install-sensu) instead.
```

The sandbox startup process takes about 5 minutes.

_NOTE: The sandbox configures VirtualBox to forward TCP ports 3002 and 4002 from the sandbox virtual machine (VM) to the localhost to make it easier for you to interact with the sandbox dashboards. Dashboard links provided in this tutorial assume port forwarding from the VM to the host is active._

**4. SSH into the sandbox**

> Thanks for waiting! To start shell access to the sandbox:

Reviewer 2: This reads sorta odd to me. Maybe `To start shell into the sandbox`?  Or `To start your shell access to the sandbox`.

> Thanks for waiting! To start shell access to the sandbox:

Author: ```suggestion
Thanks for waiting! To start shell into the sandbox:
```

{{< highlight shell >}}
vagrant ssh
{{< /highlight >}}

You should be greeted with this prompt:  

{{< highlight shell >}}
[sensu_go_sandbox]$
{{< /highlight >}}

> - To exit out of the sandbox, press `CTRL`+`D`.

Reviewer: +1 to these verbs over "use"! way more descriptive 👍

- To erase and restart the sandbox, run `vagrant destroy` and then `vagrant up`.
- To reset the sandbox's Sensu configuration to the beginning of this tutorial, run `vagrant provision`.

> _NOTE: The sandbox pre-configures sensuctl with the Sensu Go admin user, so you won't have to configure sensuctl each time you spin up the sandbox to try out a new feature. Before installing sensuctl outside of the sandbox, read the [first time setup reference][1] to learn how to configure sensuctl._  

Reviewer 2: For some reason the link to `[1]` still throws me to the 5.0 release (the original link). I do see that isn't the case down below, so might be a cache thing for me. If no one else has that issue, let's not worry about it.

Reviewer 2: 🤦‍♂ . I had the wrong branch checked out locally. Disregard.

agent      system:agents    true    
{{< /highlight >}}

Sensu keeps track of monitored components as entities. Start by using sensuctl to make sure Sensu hasn't connected to any entities yet:

{{< highlight shell >}}
sensuctl entity list
 ID   Class   OS   Subscriptions   Last Seen  

> ──── ─────── ──── ─────────────── ─────────── 

Author: The second time I ran through the sandbox (and subsequent times), the entity list includes sensu-go-sandbox (in other words, the initial `sensuctl entity list` isn't empty). I think we should add a cleanup/delete command at the end of this sandbox lesson.

Reviewer 2: Most likely stuff that remains in `etcd`. As the lesson is meant to be done from start to finish, I think having them blow away the VM is fine. Secondary runs will work quicker as they won't have to re-download the base image of the OS.

So `vagrant destroy`.

Author: Thank you! I kept trying to do it with sensuctl and sensuctl was like "idk what you want 🤷‍♀ "

{{< /highlight >}}

> Now you can start the Sensu agent to start monitoring the sandbox:

Reviewer 2: ```suggestion
Now you can start the Sensu agent to begin monitoring the sandbox:
```

I have no reason other than start is used twice closely together. Disregarding this suggestion is good as well.

> sensu-go-sandbox   keepalive   Keepalive last sent from sensu-go-sandbox at 2019-01-24 21:29:06 +0000 UTC        0   false      2019-01-24 21:29:06 +0000 UTC 

Author: Here again, on second and subsequent workthroughs, this list includes `check_disk_usage` and `curl_timings` in addition to the keepalive event.

Reviewer 2: See https://github.com/sensu/sensu-docs/pull/1833#discussion_r336226957, let's destroy the VM at the end of the lesson. I don't believe it was ever designed to be ran again with the current VM resource.

{{< /highlight >}}

The sensu-go-sandbox keepalive event has status 0, which means the agent is in an OK state and is able to communicate with the Sensu backend.

You can also see the event and the entity in the [Sensu dashboard](http://localhost:3002). Log in to the dashboard with the default admin credentials: username `admin` and password `P@ssw0rd!`.

## Lesson \#2: Pipe keepalive events into Slack

Now that you know the sandbox is working properly, let's get to the fun stuff: creating a workflow. In this lesson, you'll create a workflow that sends keepalive alerts to Slack.

> _**NOTE**: If you'd rather not create a Slack account, you skip ahead to [Lesson \#3](#lesson-3-automate-event-production-with-the-sensu-agent)._

Reviewer: I'd keep "can" here -- "you can skip ahead" sounds a little more friendly than "you skip ahead"

> _**NOTE**: If you'd rather not create a Slack account, you skip ahead to [Lesson \#3](#lesson-3-automate-event-production-with-the-sensu-agent)._

Author: ```suggestion
_**NOTE**: If you'd rather not create a Slack account, you can skip ahead to [Lesson \#3](#lesson-3-automate-event-production-with-the-sensu-agent)._
```

**1. Get your Slack webhook URL**

[Create a Slack workspace][2] (or use an existing workspace, if you’re already a Slack admin).

> Then, visit `YOUR-WORKSPACE-NAME.slack.com/services/new/incoming-webhook`. Follow the steps to add the *Incoming WebHooks* integration and save your webhook. Your webhook channel and URL will be listed under Integration Settings--you’ll need both later in this lesson.

Reviewer: same comment as above about the em dash

> Then, visit `YOUR-WORKSPACE-NAME.slack.com/services/new/incoming-webhook`. Follow the steps to add the *Incoming WebHooks* integration and save your webhook. Your webhook channel and URL will be listed under Integration Settings--you’ll need both later in this lesson.

Author: ```suggestion
Then, visit `YOUR-WORKSPACE-NAME.slack.com/services/new/incoming-webhook`. Follow the steps to add the *Incoming WebHooks* integration and save your webhook. Your webhook channel and URL will be listed under Integration Settings &mdash; you’ll need both later in this lesson.
```

**2. Register the Sensu Slack handler asset**

[Assets][3] are shareable, reusable packages that make it easy to deploy Sensu plugins. In this lesson, we'll use the [Sensu Slack handler asset][4] to power a `slack` handler.

Use sensuctl to register the [Sensu Slack handler asset][4].

{{< highlight shell >}}

> sensuctl asset create sensu-slack-handler --url "https://assets.bonsai.sensu.io/3149de09525d5e042a83edbb6eb46152b02b5a65/sensu-slack-handler_1.0.3_linux_amd64.tar.gz" --sha512 "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8"

Author: When I run
```
sensuctl asset create sensu-slack-handler --url "https://assets.bonsai.sensu.io/3149de09525d5e042a83edbb6eb46152b02b5a65/sensu-slack-handler_1.0.3_linux_amd64.tar.gz" --sha512 "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8"
```

I get `Error: resource already exists` (is this somehow happening because I ran through the "Learn Sensu in 15 minutes" Katacoda scenario? Or is it another issue?)

Reviewer 2: Attempting to reproduce this, having some issues with vagrant at the moment. I'll come back to this.

Reviewer 2: I do not get the same error, so I would assume you already had an asset named `sensu-slack-handler` and sensuctl gave the error that it can't make it as something named that already existed:

```
[vagrant@sensu-centos ~]$ sensuctl asset create sensu-slack-handler --url "https://assets.bonsai.sensu.io/3149de09525d5e042a83edbb6eb46152b02b5a65/sensu-slack-handler_1.0.3_linux_amd64.tar.gz" --sha512 "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8"
Created
```

Author: I'm sure you're correct. It's probably from when Cameron walked me through the website quickstart...I don't know why I thought the Katacoda in-browser tutorial would have done it. Thank you!

Created
{{< /highlight >}}

The `sensu-slack-handler` asset is now ready to use with Sensu.  Use sensuctl to see the complete asset definition.

{{< highlight shell >}}
sensuctl asset info sensu-slack-handler --format yaml
{{< /highlight >}}

_PRO TIP: You can use resource definitions to create and update resources (like assets) using `sensuctl create --file filename.yaml`. See the [sensuctl docs][5] for more information._

**3. Create a Sensu Slack handler**

Open the `sensu-slack-handler.json` handler definition provided with the sandbox in your preferred text editor. Edit the definition to include your Slack channel, webhook URL, and the `sensu-slack-handler` asset.

> 

Author: ```suggestion
_**NOTE**: If you aren't sure how to open the handler and edit the definition, try these [Vi/Vim gist instructions][11]._

```

Created
{{< /highlight >}}

The `sensu-slack-handler` asset is now ready to use with Sensu.  Use sensuctl to see the complete asset definition.

{{< highlight shell >}}
sensuctl asset info sensu-slack-handler --format yaml
{{< /highlight >}}

_PRO TIP: You can use resource definitions to create and update resources (like assets) using `sensuctl create --file filename.yaml`. See the [sensuctl docs][5] for more information._

**3. Create a Sensu Slack handler**

Open the `sensu-slack-handler.json` handler definition provided with the sandbox in your preferred text editor. Edit the definition to include your Slack channel, webhook URL, and the `sensu-slack-handler` asset.

> _**NOTE**: If you aren't sure how to open the handler and edit the definition, try these [Vi/Vim gist instructions][11]._

Author: ```suggestion

_**NOTE**: If you aren't sure how to open the handler and edit the definition, try these [Vi/Vim gist instructions][11]._
```

Created
{{< /highlight >}}

The `sensu-slack-handler` asset is now ready to use with Sensu.  Use sensuctl to see the complete asset definition.

{{< highlight shell >}}
sensuctl asset info sensu-slack-handler --format yaml
{{< /highlight >}}

_PRO TIP: You can use resource definitions to create and update resources (like assets) using `sensuctl create --file filename.yaml`. See the [sensuctl docs][5] for more information._

**3. Create a Sensu Slack handler**

> Open the `sensu-slack-handler.json` handler definition provided with the sandbox, and edit the definition to include your Slack channel, webhook URL, and the `sensu-slack-handler` asset.

Author: I think we should somehow incorporate instructions for handling this in vim. Edgar walked me through it, and I drafted the rough instructions listed below as result of his help.

I would like to include these instructions right in this article (and with this step), if no one objects. If anyone thinks it's too disruptive to the flow, then I think we could add a "footnote" (just a section at the bottom of the page) -- add a link here that lets people hop down to that footnote, then a link after the instructions so they can hop back to their place.

`vi sensu-slack-handler.json` (opens json file you need to edit in Vim)
`i` (to be able to write over text)
Use arrows to navigate. No copy/paste; delete and retype:
 - Replace KEEPALIVE_SLACK_WEBHOOK value with your webhook URL
 - Replace KEEPALIVE_SLACK_CHANNEL value with your slack channel
 - In the runtime_assets [], type `"sensu-slack-handler"` (including quotation marks)

Press esc
Type `:w` (to write/save your changes)
Type `:q` (to quit back to sensuctl)

Reviewer 2: While I like having more instructions for users, this would assume they are using `vim` when they can be using `nano`, some other text editor or even be using Windows. I sorta lean towards a tool like `sed` for this, but that assumes still that they are on a Linux based system (and even given that there are GNU vs non-GNU differences). Maybe we could just add more English like `Open the handler definition in your favorite text editor`.

Author: I added the clarification as you suggested. This came up because I didn't know what to do at this point, and Edgar walked me through using vim (I think b/c it was the easiest thing to use at the time). Edgar mentioned this is something a typical sysadmin would know how to do.

What do you think of something like...adding the vim instructions as a gist that we can link to from here? I think the instructions are too detailed to insert right in the text, but I would not have been able to continue the lesson if Edgar hadn't been available to help me figure it out.

Reviewer 2: If we really want to have instructions for just `vim`, we would need to at least call out "vim or vi" as not everyone uses `vim` (vim=vi improved). I for one do not use it on OSX cause of some odd unpredictable behavior.

However the rest of the actions should be the same. A gist sounds fine to keep the guide cleaner.

Author: Would it be better to give instructions for a different text editor? I really just want to get a totally new user through this step...the only reason I have vim listed is because that's how Edgar showed me (and you can just do it within the rest of the workflow)

Author: Also maybe Edgar actually showed me in `vi`

Reviewer 2: I think `vi` and `vim` is fine for now. If we get feedback about people getting stuck, we can add a different editor or method.

Any other tool we pick or add is going to have the same problem; they might not have it. Having at least one example is a good idea.

> Open the `sensu-slack-handler.json` handler definition provided with the sandbox, and edit the definition to include your Slack channel, webhook URL, and the `sensu-slack-handler` asset.

Author: ```suggestion
Open the `sensu-slack-handler.json` handler definition provided with the sandbox in your preferred text editor. Edit the definition to include your Slack channel, webhook URL, and the `sensu-slack-handler` asset.
```

{{< highlight shell >}}
"env_vars": [

>   "KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/ABC",

Reviewer 2: So I think the point of `AAA/BBB/CCC` was that Slack webhooks follow the form of `someString/someOtherString/andYetAnotherString`. For example: `https://hooks.slack.com/services/K04L63BU1/BTZ7FGFW/WwZjiHpaeDDiXmk3QZUzlrkO`. That said, I don't think it is too important, just an FYI. If a user has a webhook URL, they'll use the entire thing.

Author: No, I think you're correct. I will change it back.

>   "KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/ABC",

Author: ```suggestion
  "KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/AAA/BBB/CCC",
```

  "KEEPALIVE_SLACK_CHANNEL=#monitoring"
],
"runtime_assets": ["sensu-slack-handler"]
{{< /highlight >}}

Now you can create a Slack handler named `keepalive` to process keepalive events.

{{< highlight shell >}}
sensuctl create --file sensu-slack-handler.json
{{< /highlight >}}

> Use sensuctl to see available event handlers--in this case, you'll only see the `keepalive` handler you just created..

Reviewer: same comment re: em dash

─────────── ────── ───────── ───────── ───────── ────────────────────────────────────────────────────────────────────────────────────────────────────────── ────────────────────────────────────────────────────────────────────────────────────────────────── ───────────────────── 
 keepalive   pipe         0                       RUN:  /usr/local/bin/sensu-slack-handler -c "${KEEPALIVE_SLACK_CHANNEL}" -w "${KEEPALIVE_SLACK_WEBHOOK}"   KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/ABC,KEEPALIVE_SLACK_CHANNEL=#monitoring   sensu-slack-handler  
{{< /highlight >}}

Sensu monitoring events should begin arriving in your Slack channel, indicating that the sandbox entity is in an OK state.

**4. Filter keepalive events**

Now that you're generating Slack alerts, let's reduce the potential for alert fatigue by adding a filter that sends only warning, critical, and resolution alerts to Slack.

To accomplish this, you'll interactively add the built-in is_incident filter to the keepalive handler, which will make sure you only receive alerts when the sandbox entity fails to send a keepalive event.

{{< highlight shell >}}
sensuctl handler update keepalive
{{< /highlight >}}

> The first prompt will be for environment variables. Just press `return` to continue. The second prompt is for the filters selection--enter `is_incident` to apply the is_incident filter.

Reviewer: ditto comment about em dash 😛

> 

Author: For info about why I made these changes, see https://github.com/sensu/sensu-docs/issues/1808#issuecomment-538073373

{{< highlight shell >}}
? Filters: is_incident
{{< /highlight >}}

> For the mutator, timeout, type, runtime assets, and command prompts, just press `return` again.

Reviewer 2: ```suggestion
For the mutator, timeout, type, runtime assets, and command prompts, just press `return` for each of them.
```

Reviewer 2: Or maybe `for each prompt`?

Author: What do you think of:
"For each of the mutator, timeout, type, runtime assets, and command prompts, just press `return`."

Reviewer 2: That works!

Use sensuctl to confirm that the keepalive handler now includes the is_incident filter:

{{< highlight shell >}}
[sensu_go_sandbox]$ sensuctl handler info keepalive
=== keepalive
Name:                  keepalive
Type:                  pipe
Timeout:               0
Filters:               is_incident
Mutator:               
Execute:               RUN:  sensu-slack-handler -c "${KEEPALIVE_SLACK_CHANNEL}" -w "${KEEPALIVE_SLACK_WEBHOOK}"
Environment Variables: KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/ABC, KEEPALIVE_SLACK_CHANNEL=#monitoring
Runtime Assets:        sensu-slack-handler
{{< /highlight >}}

With the filter in place, you should no longer receive messages in your Slack channel every time the sandbox entity sends a keepalive event.

> Let's stop the agent and confirm that your receive the expected warning message.

Reviewer: "you" receive (instead of "your")

{{< highlight shell >}}
sudo systemctl stop sensu-agent
{{< /highlight >}}

After a couple minutes, you should see a warning message in your Slack channel informing you that the sandbox entity is no longer sending keepalive events.

> Start the agent to resolve the warning.

Reviewer 2: So there might be another enhancement here. I'll have to run through the sandbox to confirm, but it is possible that applying only the `is_incident` filter without the `is_resolution` one will only allow warns, critical or unknown alerts through and not the resolution. Up on line 187 it is mentioned that resolutions will come through, so I think I'll defer that away from this PR and confirm it once this is merged and I've gone through the entire thing.

Author: OK, thank you!

Reviewer 2: I confirmed resolutions come through. Resolving.

"runtime_assets": ["sensu-slack-handler"]
{{< /highlight >}}

Now you can create a Slack handler named `keepalive` to process keepalive events.

{{< highlight shell >}}
sensuctl create --file sensu-slack-handler.json
{{< /highlight >}}

> Use sensuctl to see available event handlers--in this case, you'll only see the `keepalive` handler you just created..

Author: ```suggestion
Use sensuctl to see available event handlers &mdash; in this case, you'll only see the `keepalive` handler you just created..
```

{{< highlight shell >}}
sensuctl handler list
  Name      Type   Timeout   Filters   Mutator                                                   Execute                                                                                                              Environment Variables                            Assets         
─────────── ────── ───────── ───────── ───────── ────────────────────────────────────────────────────────────────────────────────────────────────────────── ────────────────────────────────────────────────────────────────────────────────────────────────── ───────────────────── 

>  keepalive   pipe         0                       RUN:  /usr/local/bin/sensu-slack-handler -c "${KEEPALIVE_SLACK_CHANNEL}" -w "${KEEPALIVE_SLACK_WEBHOOK}"   KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/ABC,KEEPALIVE_SLACK_CHANNEL=#monitoring   sensu-slack-handler  

Author: ```suggestion
 keepalive   pipe         0                       RUN:  /usr/local/bin/sensu-slack-handler -c "${KEEPALIVE_SLACK_CHANNEL}" -w "${KEEPALIVE_SLACK_WEBHOOK}"   KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/AAA/BBB/CCC,KEEPALIVE_SLACK_CHANNEL=#monitoring   sensu-slack-handler  
```

{{< /highlight >}}

Sensu monitoring events should begin arriving in your Slack channel, indicating that the sandbox entity is in an OK state.

**4. Filter keepalive events**

Now that you're generating Slack alerts, let's reduce the potential for alert fatigue by adding a filter that sends only warning, critical, and resolution alerts to Slack.

To accomplish this, you'll interactively add the built-in is_incident filter to the keepalive handler, which will make sure you only receive alerts when the sandbox entity fails to send a keepalive event.

{{< highlight shell >}}
sensuctl handler update keepalive
{{< /highlight >}}

> The first prompt will be for environment variables. Just press `return` to continue. The second prompt is for the filters selection--enter `is_incident` to apply the is_incident filter.

Author: ```suggestion
The first prompt will be for environment variables. Just press `return` to continue. The second prompt is for the filters selection &mdash; enter `is_incident` to apply the is_incident filter.
```

{{< highlight shell >}}
? Filters: is_incident
{{< /highlight >}}

> For the mutator, timeout, type, runtime assets, and command prompts, just press `return` again.

Author: ```suggestion
For each of the mutator, timeout, type, runtime assets, and command prompts, just press `return`.
```

Use sensuctl to confirm that the keepalive handler now includes the is_incident filter:

{{< highlight shell >}}
[sensu_go_sandbox]$ sensuctl handler info keepalive
=== keepalive
Name:                  keepalive
Type:                  pipe
Timeout:               0
Filters:               is_incident
Mutator:               
Execute:               RUN:  sensu-slack-handler -c "${KEEPALIVE_SLACK_CHANNEL}" -w "${KEEPALIVE_SLACK_WEBHOOK}"

> Environment Variables: KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/ABC, KEEPALIVE_SLACK_CHANNEL=#monitoring

Author: ```suggestion
Environment Variables: KEEPALIVE_SLACK_WEBHOOK=https://hooks.slack.com/services/AAA/BBB/CCC, KEEPALIVE_SLACK_CHANNEL=#monitoring
```

Runtime Assets:        sensu-slack-handler
{{< /highlight >}}

With the filter in place, you should no longer receive messages in your Slack channel every time the sandbox entity sends a keepalive event.

> Let's stop the agent and confirm that your receive the expected warning message.

Author: ```suggestion
Let's stop the agent and confirm that you receive the expected warning message.
```

**2. Install Nginx and the Sensu HTTP Plugin**

You'll use the [Sensu HTTP Plugin](https://github.com/sensu-plugins/sensu-plugins-http) to monitor an Nginx server running on the sandbox.

First, install and start Nginx:

{{< highlight shell >}}

> sudo yum install -y nginx && sudo systemctl start nginx

Reviewer 2: Unrelated to this PR, this command to install nginx didn't work for me. It might be due to me having to use CentOS 7.4, but not sure. Should confirm with another user.

Author: Cameron: need to run before `yum install -y nginx`
`yum install -y epel-release`

Reviewer 2: Confirmed installing EPEL resolved the issue.

**2. Install Nginx and the Sensu HTTP Plugin**

You'll use the [Sensu HTTP Plugin](https://github.com/sensu-plugins/sensu-plugins-http) to monitor an Nginx server running on the sandbox.

> First, install and start Nginx:

Author: ```suggestion
First, install the EPEL release package:
`yum install -y epel-release`

Then, install and start Nginx:
```

**2. Install Nginx and the Sensu HTTP Plugin**

You'll use the [Sensu HTTP Plugin](https://github.com/sensu-plugins/sensu-plugins-http) to monitor an Nginx server running on the sandbox.

First, install the EPEL release package:

> `yum install -y epel-release`

Author: ```suggestion
{{< highlight shell >}}
yum install -y epel-release
{{< /highlight >}}
```

**2. Install Nginx and the Sensu HTTP Plugin**

You'll use the [Sensu HTTP Plugin](https://github.com/sensu-plugins/sensu-plugins-http) to monitor an Nginx server running on the sandbox.

First, install the EPEL release package:

> {{< highlight shell >}}

Author: ```suggestion

{{< highlight shell >}}
```

**2. Install Nginx and the Sensu HTTP Plugin**

You'll use the [Sensu HTTP Plugin](https://github.com/sensu-plugins/sensu-plugins-http) to monitor an Nginx server running on the sandbox.

First, install the EPEL release package:

{{< highlight shell >}}

> yum install -y epel-release

Author: ```suggestion
sudo yum install -y epel-release
```

**2. Install Nginx and the Sensu HTTP Plugin**

You'll use the [Sensu HTTP Plugin](https://github.com/sensu-plugins/sensu-plugins-http) to monitor an Nginx server running on the sandbox.

First, install the EPEL release package:

{{< highlight shell >}}
sudo yum install -y epel-release
{{< /highlight >}}

Then, install and start Nginx:

{{< highlight shell >}}
sudo yum install -y nginx && sudo systemctl start nginx
{{< /highlight >}}

Make sure it's working:

{{< highlight shell >}}

> curl -I http://localhost:80

Author: ```suggestion
curl -I http://localhost:80
...
```

**2. Install Nginx and the Sensu HTTP Plugin**

You'll use the [Sensu HTTP Plugin](https://github.com/sensu-plugins/sensu-plugins-http) to monitor an Nginx server running on the sandbox.

First, install the EPEL release package:

{{< highlight shell >}}
sudo yum install -y epel-release
{{< /highlight >}}

Then, install and start Nginx:

{{< highlight shell >}}
sudo yum install -y nginx && sudo systemctl start nginx
{{< /highlight >}}

Make sure it's working:

{{< highlight shell >}}
curl -I http://localhost:80

> ...

Author: ```suggestion
```

{{< /highlight >}}

{{< highlight shell >}}

> HTTP/1.1 200 OK

Author: ```suggestion
HTTP/1.1 200 OK
...
```

"runtime_assets": ["sensu-influxdb-handler"]
{{< /highlight >}}

Now you can use sensuctl to create the `influx-db` handler:

{{< highlight shell >}}
sensuctl create --file influx-handler.json
{{< /highlight >}}

Use sensuctl to confirm that the handler was created successfully.

{{< highlight shell >}}
sensuctl handler list
{{< /highlight >}}

The `influx-db` handler should be listed. If you completed [lesson \#2](#lesson-2-pipe-keepalive-events-into-slack), you'll also see the `keepalive` handler.

**4. Create a check to monitor Nginx**

> 

Author: The instructions here do not tell you where to type `metrics-curl.rb` (I guessed that it's a handler based on the info below, and my response for `sensuctl create --file curl_timings-check.json` matches the one in the lesson, but I'm still not sure I did this step correctly).

Reviewer 3: `metrics-curl.rb` is a check plugin which is referenced in the value of the `command` field in `curl_timings-check.json`. See https://github.com/sensu/sandbox/blob/master/sensu-go/files/curl_timings-check.json#L10

Can we make the connection more clear?

Author: OK -- so you don't have to open `curl_timings-check.json` at all in this step?

Author: I'm thinking something like...

Use the `curl_timings-check.json` file provided with the sandbox to create a service check that runs the `metrics-curl.rb` check plugin every 10 seconds on all entities with the `entity:sensu-go-sandbox` subscription and sends events to the InfluxDB pipeline. The `metrics-curl.rb` plugin is already included as the value of the `command` field in `curl_timings-check.json`.

curl_timings   /opt/sensu-plugins-ruby/embedded/bin/metrics-curl.rb -u "http://localhost"         10                0     0   entity:sensu-go-sandbox                               true       false    graphite_plaintext   influx-db        
{{< /highlight >}}

> This check defines a metrics handler and metric format. In Sensu Go, metrics are a core element of the data model: you can build pipelines to handle metrics separately from alerts. This allows you to customize your monitoring workflows to get better visibility and reduce alert fatigue.

Reviewer 3: ```suggestion
This check specifies a metrics handler and metric format. In Sensu Go, metrics are a core element of the data model: you can build pipelines to handle metrics separately from alerts. This allows you to customize your monitoring workflows to get better visibility and reduce alert fatigue.
```

sensuctl check list
{{< /highlight >}}

{{< highlight shell >}}
     Name                                        Command                                     Interval   Cron   Timeout   TTL        Subscriptions        Handlers   Assets   Hooks   Publish?   Stdin?     Metric Format      Metric Handlers  
────────────── ──────────────────────────────────────────────────────────────────────────── ────────── ────── ───────── ───── ───────────────────────── ────────── ──────── ─────── ────────── ──────── ──────────────────── ───────────────── 
curl_timings   /opt/sensu-plugins-ruby/embedded/bin/metrics-curl.rb -u "http://localhost"         10                0     0   entity:sensu-go-sandbox                               true       false    graphite_plaintext   influx-db        
{{< /highlight >}}

This check specifies a metrics handler and metric format. In Sensu Go, metrics are a core element of the data model: you can build pipelines to handle metrics separately from alerts. This allows you to customize your monitoring workflows to get better visibility and reduce alert fatigue.

After about 10 seconds, you can see the event produced by the entity:

{{< highlight shell >}}

> sensuctl event info sensu-go-sandbox curl_timings --format json | jq .

Author: When I run:
`sensuctl event info sensu-go-sandbox curl_timings --format json | jq .`

I get:
`-bash: jq: command not found`

Author: When I run:
`sensuctl event info sensu-go-sandbox curl_timings --format json` (without the ` | jq.`), I get the response shown (`"history`" etc.)

Reviewer 3: The `epel-release` and `jq` packages are meant to be installed in https://github.com/sensu/sandbox/blob/master/sensu-go/provision/setup.sh#L79-L80 and https://github.com/sensu/sandbox/blob/master/sensu-go/provision/setup.sh#L99 respectively. I suspect the logic to determine whether `epel-release` is needed is failing us somehow.

Author: Yay it works!

...
{{< /highlight >}}

> Because you configured a metric format, the Sensu agent can convert the Graphite-formatted metrics provided by the check command into a set of Sensu-formatted metrics (not shown in the output). These metrics are then sent to the InfluxDB handler, which reads Sensu-formatted metrics and converts them to a format InfluxDB accepts.

Reviewer 3: ```suggestion
Because the check definition specified a metric format of `graphite_plaintext`, the Sensu agent will treat the output of the check command as Graphite-formatted metrics and translate them into a set of Sensu-formatted metrics (not shown in the output). These metrics are then sent to the InfluxDB handler, which reads Sensu-formatted metrics and converts them to a format InfluxDB accepts.
```

Author: Yes, much better, thank you!

_**NOTE**: Metric support isn't limited to Graphite! The Sensu agent can extract metrics in multiple line protocol formats, including Nagios performance data._

**5. See the HTTP response code events for Nginx in Grafana.**

> Log in to [Grafana](http://localhost:4002/d/go01/sensu-go-sandbox) with username: `admin` and password: `admin`. You should see a graph of live HTTP response codes for Nginx.

Author: Grafana prompts you to change your password when you follow this link (you can skip this, but it seems kind of weird and disrupts the stream of the training). Any way we can link to a page that's already logged in?

Reviewer 2: We might be able to automate the changing of the password on the backend. I can help out with that in this PR or if you want to get this merged and we can iterate later, that works too!

Author: It would be great if we can just have them click a link that opens the Sensu Go sandbox page in Grafana that automatically logs them in...but if we can't do that, it would at least be much better if we can work around the password change issue! I agree this might be better for a follow-up PR.

Reviewer 2: I do believe that Grafana has a read only user that doesn't require a password, so maybe we can leverage that as we aren't having them change anything inside Grafana.

Reviewer 3: I think we can make some changes in the sandbox repo so that a login is not necessary. See https://stackoverflow.com/questions/33111835/how-to-set-up-grafana-so-that-no-password-is-necessary-to-view-dashboards for some relevant discussion.

Author: Awesome! Cameron is this something you can help me learn to do? I think I should learn how to do it so I have it in my pocket for future scenarios.

...
{{< /highlight >}}

Because the check definition specified a metric format of `graphite_plaintext`, the Sensu agent will treat the output of the check command as Graphite-formatted metrics and translate them into a set of Sensu-formatted metrics (not shown in the output). These metrics are then sent to the InfluxDB handler, which reads Sensu-formatted metrics and converts them to a format InfluxDB accepts.

_**NOTE**: Metric support isn't limited to Graphite! The Sensu agent can extract metrics in multiple line protocol formats, including Nagios performance data._

**5. See the HTTP response code events for Nginx in Grafana.**

> Log in to [Grafana](http://localhost:4002/d/go01/sensu-go-sandbox) with username: `admin` and password: `admin`. You should see a graph of live HTTP response codes for Nginx.

Author: You still have to log in with the `admin` credentials, and Grafana still prompts you to change your password (you can click Skip)

Reviewer 3: Is this the case with the `anonymous-grafana` branch for https://github.com/sensu/sandbox/pull/60 checked out? That PR needs to be merged before the grafana changes are expected to take effect on `master` branch.

Author: I forgot we hadn't merged that one -- I tried again with that branch checked out. With the branch checked out, Grafana won't load. I made a short gif to show what I see:
![2019-10-21_20-11-37](https://user-images.githubusercontent.com/19190208/67252060-5e140c80-f43f-11e9-82d9-a962d3eff6c7.gif)

Reviewer 2: It actually does work, it's just listening on the wrong port (3000) and also caused the Sensu Dashboard to not work either.

The setup script actually changes the default port from 3000 to 4000 to not conflict with the Sensu Go Dashboard here:

https://github.com/sensu/sandbox/blob/master/sensu-go/provision/setup.sh#L140-L141

But that is lost with the addition of the file in the PR https://github.com/sensu/sandbox/pull/60 as the recursive coping of the files in `/vagrant_files/etc/grafana/` clobbers the one created by the installation of Grafana and modified by the `sed` command mentioned above. We should add the custom port line of `http_port = 4000` to the just mentioned PR and remove it from setup.sh. It is the only customization I see in setup.sh against the default config.

Also in setup.sh, line 140 has a reference to Uchiwa when that should be the Sensu Go Dashboard.

Reviewer 2: And with all these changes, we should consider a mention somewhere in this docs page that previous users should do a `git pull` to get the latest updates, as we are making other changes that assume a user may re-run this at a later time.

sensuctl create --file disk_usage-check.json
{{< /highlight >}}

> _**NOTE**: You don't need to make any changes to disk_usage-check.json before running `sensuctl create --file disk_usage-check.json`._

Reviewer 2: We need to rename the `disk_usage-check.json` in the sandbox lesson if we want to make this line in a note. The current `_` after the word `disk` is causing the note to not be rendered correctly in browser.

Author: Thank you! I'll fix that (remove the note formatting)

sensuctl create --file disk_usage-check.json
{{< /highlight >}}

_**NOTE**: You don't need to make any changes to disk_usage-check.json before running `sensuctl create --file disk_usage-check.json`._

> You should see the check working in the [dashboard entity view](http://localhost:3002/#/entities) and via sensuctl:

Reviewer: should we be using https for these dashboard links?

Reviewer 2: While `https` is typically preferred for anything that traverses a public network (the Internet), we don't need it for this lesson. All this traffic is via localhost (the user's computer) so any worry about non-encrypted traffic is low (if their computer is compromised, a Slack URL is the least of their worries). It would also require some non-default configuration changes on the Sensu Go end as it defaults to `http`, along with requiring a certificate or the user would get a warning about not being able to trust the site in their browser. Good question though!

{{< highlight shell >}}
sensuctl event list
{{< /highlight >}}

Now, you should be able to see disk usage metrics for the sandbox in Grafana: refresh your Grafana tab or [reload with this link](http://localhost:4002/d/go02/sensu-go-sandbox-combined).

You made it! You're ready for the next level of Sensu-ing. Here are some resources to help continue your journey:

- Try another [lesson in the Sensu sandbox][7]
- [Install Sensu Go][8]
- [Collect StatsD metrics][9]

> - [Create a ready-only user][10]

Author: Maybe we should end this lesson with a command that deletes all the resources we created (a la https://docs.sensu.io/sensu-go/5.13/sensuctl/reference/#deleting-resources) -- or will subsequent lessons rely on these resources in some way?

Reviewer 2: I don't believe our lessons, at least in the sandbox, depend on previous resources. So that would be a good add I think, but it would probably require some additions to the other guides like "If you did this lesson, you can skip these things...". Probably out of scope for this PR.

Author: Something I can look into in a follow-up...

Reviewer 2: I think we can resolve this section with having a `vagrant destroy` as suggested for:

https://github.com/sensu/sensu-docs/pull/1833#discussion_r336226957

We actually don't clean up the VM at all, so new users of Vagrant might be surprised to find that VM running later on.

{{< /highlight >}}

_**NOTE**: You don't need to make any changes to disk_usage-check.json before running `sensuctl create --file disk_usage-check.json`._

You should see the check working in the [dashboard entity view](http://localhost:3002/#/entities) and via sensuctl:

{{< highlight shell >}}
sensuctl event list
{{< /highlight >}}

Now, you should be able to see disk usage metrics for the sandbox in Grafana: refresh your Grafana tab or [reload with this link](http://localhost:4002/d/go02/sensu-go-sandbox-combined).

You made it! You're ready for the next level of Sensu-ing. Here are some resources to help continue your journey:

- Try another [lesson in the Sensu sandbox][7]
- [Install Sensu Go][8]
- [Collect StatsD metrics][9]
- [Create a ready-only user][10]

[1]: ../../sensuctl/reference/#first-time-setup
[2]: https://slack.com/get-started#create
[3]: ../../reference/assets
[4]: https://bonsai.sensu.io/assets/sensu/sensu-slack-handler
[5]: ../../sensuctl/reference#creating-resources
[6]: https://bonsai.sensu.io/assets/sensu/sensu-influxdb-handler
[7]: ../../getting-started/sandbox/
[8]: ../../installation/install-sensu
[9]: ../../guides/aggregate-metrics-statsd

> [10]: ../../guides/create-read-only-user/

Author: ```suggestion
[10]: ../../guides/create-read-only-user/
[11]: https://gist.github.com/hillaryfraley/838a046821171b1a37d0dafb16584518
```

sensuctl create --file disk_usage-check.json
{{< /highlight >}}

> _**NOTE**: You don't need to make any changes to disk_usage-check.json before running `sensuctl create --file disk_usage-check.json`._

Author: ```suggestion
You don't need to make any changes to disk_usage-check.json before running `sensuctl create --file disk_usage-check.json`.
```

You should see the check working in the [dashboard entity view](http://localhost:3002/#/entities) and via sensuctl:

{{< highlight shell >}}
sensuctl event list
{{< /highlight >}}

Now, you should be able to see disk usage metrics for the sandbox in Grafana: refresh your Grafana tab or [reload with this link](http://localhost:4002/d/go02/sensu-go-sandbox-combined).

> You made it! You're ready for the next level of Sensu-ing. Here are some resources to help continue your journey:

Author: ```suggestion
You made it! You're ready for the next level of Sensu-ing.

Before you move on, take a moment to remove the virtual machine and resources installed during this sandbox lesson. Run:
{{< highlight shell >}}
vagrant destroy
{{< /highlight >}}

Now you can continue exploring Sensu with a clean slate. Here are some resources to help continue your journey:
```

sensuctl create --file disk_usage-check.json
{{< /highlight >}}

You don't need to make any changes to disk_usage-check.json before running `sensuctl create --file disk_usage-check.json`.

You should see the check working in the [dashboard entity view](http://localhost:3002/#/entities) and via sensuctl:

{{< highlight shell >}}
sensuctl event list
{{< /highlight >}}

> Now, you should be able to see disk usage metrics for the sandbox in Grafana: refresh your Grafana tab or [reload with this link](http://localhost:4002/d/go02/sensu-go-sandbox-combined).

Author: ```suggestion
Now, you should be able to see disk usage metrics for the sandbox in Grafana: [reload your Grafana tab to show the Sensu Go Sandbox Combined](http://localhost:4002/d/go02/sensu-go-sandbox-combined).
```

You made it! You're ready for the next level of Sensu-ing.

> Before you move on, take a moment to remove the virtual machine and resources installed during this sandbox lesson. Run:

Author: ```suggestion
Before you move on, take a moment to remove the virtual machine and resources installed during this sandbox lesson. Press `CTRL`+`D` to exit the sandbox. Then run:
```

{{< highlight shell >}}
vagrant destroy

> {{< /highlight >}}

Author: Now when I run `vagrant destroy` at this point, I get:
```
[sensu_go_sandbox]$ vagrant destroy
-bash: vagrant: command not found
```

I have gone through the lesson twice with this result (after Aaron helped me ID the previous issue). I deleted the VM manually via Virtual Box interface before going back through the lesson the second time.

Reviewer 3: Command prompt here indicates you are still inside the `sensu_go_sandbox` VM, so you need to use ctrl-D again to exit that ssh session before the `vagrant` command is available again in the context of the VM host system (your laptop).

Author: Yes, that was it! Thank you -- I had a feeling I had been looking at this for too long.

sensuctl create --file disk_usage-check.json
{{< /highlight >}}

You don't need to make any changes to disk_usage-check.json before running `sensuctl create --file disk_usage-check.json`.

You should see the check working in the [dashboard entity view](http://localhost:3002/#/entities) and via sensuctl:

{{< highlight shell >}}
sensuctl event list
{{< /highlight >}}

Now, you should be able to see disk usage metrics for the sandbox in Grafana: [reload your Grafana tab to show the Sensu Go Sandbox Combined](http://localhost:4002/d/go02/sensu-go-sandbox-combined).

You made it! You're ready for the next level of Sensu-ing.

Before you move on, take a moment to remove the virtual machine and resources installed during this sandbox lesson. Press `CTRL`+`D` to exit the sandbox. Then run:
{{< highlight shell >}}
vagrant destroy
{{< /highlight >}}

Now you can continue exploring Sensu with a clean slate. Here are some resources to help continue your journey:

- Try another [lesson in the Sensu sandbox][7]
- [Install Sensu Go][8]
- [Collect StatsD metrics][9]

> - [Create a ready-only user][10]

Author: ```suggestion
- [Create a read-only user][10]
```
